File: 0649-dota2-senate/0649-dota2-senate.dart
import 'dart:collection';

class Solution {
  String predictPartyVictory(String senate) {
    int n = senate.length;
    Queue<int> rQ = Queue<int>();
    Queue<int> dQ = Queue<int>();
    for (int i = 0; i < n; i++) {
      if (senate[i] == 'R') {
        rQ.add(i);
      } else {
        dQ.add(i);
      }
    }

    while (rQ.isNotEmpty && dQ.isNotEmpty) {
      int rIndex = rQ.removeFirst();
      int dIndex = dQ.removeFirst();

      if (rIndex < dIndex) {
        rQ.add(rIndex + n);
      } else {
        dQ.add(dIndex + n);
      }
    }

    return rQ.isEmpty ? "Dire" : "Radiant";
  }
}

File: 0649-dota2-senate/0649-dota2-senate.cpp
class Solution {
public:
    string predictPartyVictory(string senate) {
        queue<int>r;
        queue<int> d;
        int n = senate.length();
        for(int i=0;i<n;i++){
            if(senate[i]=='R')
                r.push(i);
            else 
                d.push(i);
        }
        while (!r.empty() && !d.empty()) {
            int r_index = r.front();
            int d_index = d.front();
            
            r.pop();
            d.pop();
            if (r_index < d_index) {
                r.push(r_index + n);
            } else {
                d.push(d_index + n);
            }
        }
        return r.empty()? "Dire":"Radiant";
    }
};